src/iothread.c

```diff
@@ -0,0 +1,631 @@
+    int processed = 0;
+    for (int i = 1; i < server.io_threads_num; i++) {
+        int len = listLength(mainThreadPendingClientsToIOThreads[i]);
+        if (len > 0) {
+            IOThread *t = &IOThreads[i];
+            pthread_mutex_lock(&t->pending_clients_mutex);
+            listJoin(t->pending_clients, mainThreadPendingClientsToIOThreads[i]);
+            pthread_mutex_unlock(&t->pending_clients_mutex);
+            /* Trigger an event, maybe an error is returned when buffer is full
+             * if using pipe, but no worry, io thread will handle all clients
+             * in list when receiving a notification. */
+            triggerEventNotifier(t->pending_clients_notifier);
+        }
+        processed += len;
+    }
+    return processed;
+}
+
+extern int ProcessingEventsWhileBlocked;
+
+/* The main thread processes the clients from IO threads, these clients may have
+ * a complete command to execute or need to be freed. Note that IO threads never
+ * free client since this operation access much server data.
+ *
+ * Please notice that this function may be called reentrantly, i,e, the same goes
+ * for handleClientsFromIOThread and processClientsOfAllIOThreads. For example,
+ * when processing script command, it may call processEventsWhileBlocked to
+ * process new events, if the clients with fired events from the same io thread,
+ * it may call this function reentrantly. */
+void processClientsFromIOThread(IOThread *t) {
+    listNode *node = NULL;
+
+    while (listLength(mainThreadProcessingClients[t->id])) {
+        /* Each time we pop up only the first client to process to guarantee
+         * reentrancy safety. */
+        if (node) zfree(node);
+        node = listFirst(mainThreadProcessingClients[t->id]);
+        listUnlinkNode(mainThreadProcessingClients[t->id], node);
+        client *c = listNodeValue(node);
+
+        /* Make sure the client is readable or writable in io thread to
+         * avoid data race. */
+        serverAssert(!(c->io_flags & (CLIENT_IO_READ_ENABLED | CLIENT_IO_WRITE_ENABLED)));
+        serverAssert(!(c->flags & CLIENT_CLOSE_ASAP));
+
+        /* Let main thread to run it, set running thread id first. */
+        c->running_tid = IOTHREAD_MAIN_THREAD_ID;
+
+        /* If a read error occurs, handle it in the main thread first, since we
+         * want to print logs about client information before freeing. */
+        if (c->read_error) handleClientReadError(c);
+
+        /* The client is asked to close in IO thread. */
+        if (c->io_flags & CLIENT_IO_CLOSE_ASAP) {
+            freeClient(c);
+            continue;
+        }
+
+        /* Update the client in the mem usage */
+        updateClientMemUsageAndBucket(c);
+
+        /* Process the pending command and input buffer. */
+        if (!c->read_error && c->io_flags & CLIENT_IO_PENDING_COMMAND) {
+            c->flags |= CLIENT_PENDING_COMMAND;
+            if (processPendingCommandAndInputBuffer(c) == C_ERR) {
```

Comment:
so each time we get here, we process just one command on behalf of that client from the main thread, even if there are additional commands already in the query buffer?
i.e. we then put the reply for the io thread to send and wait for the IO thread to wake us for the next command to be processed on that client (after it populated the argv array). is that right?

if it is, then i think it could be extremely inefficient in the sense even with clients that use pipeline, it will cause a `write` syscall per command, unlike before where we could write a reply of many commands at once. (this will also be inefficient on the network, causing many small TCP packets.

i think we must make an effort to parse multiple commands on behalf of each client together, then execute them together, and reply on them together.

i'd also like the same mechanism of parsing multiple commands before execution even without IO threads enabled, since it'll enable look-ahead prefetch for ROF, and solve a similar problem of many small writes there.

Reply:
> i think we must make an effort to parse multiple commands on behalf of each client together, then execute them together, and reply on them together.

yes. when we tried to do it, we found it is not easy, so we plan to optimize it in a future PR. When parsing multiple commands, we might encounter errors, for some errors, we can continue, such as `*0`; for some errors, we need to close client, such as protocol format error, but we need to process already parsed commands.

Reply:
> so currently the first command will be parsed by the IO thread, and then after the main thread executes it, it'll parse and execute the rest of the command in the query buffer, and `write` the response for all in one write?

yes.